packages/astro/src/core/middleware/callMiddleware.ts

```diff
@@ -38,13 +44,28 @@ import { AstroError, AstroErrorData } from '../errors/index.js';
 export async function callMiddleware(
 	onRequest: MiddlewareHandler,
 	apiContext: APIContext,
-	responseFunction: () => Promise<Response> | Response
+	responseFunction: (
+		apiContext: APIContext,
+		reroutePayload?: ReroutePayload
+	) => Promise<Response> | Response,
+	// TODO: remove these two arguments once rerouting goes out of experimental
+	enableRerouting: boolean,
+	logger: Logger
 ): Promise<Response> {
 	let nextCalled = false;
 	let responseFunctionPromise: Promise<Response> | Response | undefined = undefined;
-	const next: MiddlewareNext = async () => {
+	const next: MiddlewareNext = async (payload) => {
 		nextCalled = true;
-		responseFunctionPromise = responseFunction();
+		if (enableRerouting) {
+			responseFunctionPromise = responseFunction(apiContext, payload);
+		} else {
+			logger.warn(
+				'router',
+				'You tried to use the routing feature without enabling it via experimental flag. This is not allowed.'
+			);
+			responseFunctionPromise = responseFunction(apiContext);
+		}
+		// We need to pass the APIContext pass to `callMiddleware` because it can be mutated across middleware functions
```

Comment:
@bluwy These are the only code changes I added to this PR. The rest have been reviewed already or they are only docs changes.

Reply:
Thanks! Looks great to me then.